### ISSUE TITLE
[Development] Publishing Safari Extension

### ISSUE THREAD

**Issue by Aayush9029** [FreshSupaSulley/Duochrome#2]
First of all, thank you so much; you are the best.

I've shilled $100 for the developer account ;-; (I know).

I'd be more than happy and sincerely willing to help with publishing, or at the very least, notarizing or signing the extension.

I understand the complexities and nuances associated with these processes, so I'm prepared to navigate them with great care and attention. If there's anything specific you would like me to handle or any particular way you'd like me to assist, please feel free to let me know.

**UPDATE:** Clean Install + First RUN 
I've managed to compile it and, to a certain extent, bring up the UI.
Strangely, the feature activates, but the app provides an "invalid code" error. I've uploaded a comprehensive video detailing my process.

https://github.com/FreshSupaSulley/DuOSU/assets/43297314/75998deb-1cf6-4f79-a0c1-508495ae3d0f

PS: For some reason, my browser doesn’t like saving 🍪 when it comes to DUO. I doubt this is the cause of the issue, but mentioning in case.

**Comment by FreshSupaSulley** [comment]
Wow. Did not expect this to happen ;-;
I appreciate the offer! Didn't think someone would care enough to make it a real deal for Safari. I hope you spent the $100 for reasons other than DuOSU lmao

From what I've been reading, porting an extension to Safari is very easy because most of the code will transfer fine. The challenge is that I don't know much about working with Xcode, how collaboration would work between us (and if that involves Apple Developer teams), signing, etc... I imagine we will need to publish it under your name (which is fine)? If you have any insight on the details that would be great.

Also, just FYI: if we do publish it and you ever decide to stop spending the $100 for the developer fee, that's totally fine, even if they kick it off the App Store. The last thing I want to do is make you feel locked into paying for it to keep it online.

**Comment by Aayush9029** [comment]
Feel free to take a look at my updated response above.

I use the developer account for other shenanigans as well. Don't worry about the fees!

**Comment by FreshSupaSulley** [comment]
WOW, you already have it setup that's hilarious.
I also ran into a storage issue with Firefox development, that's probably what's happening

I found this solution in the Mozilla docs. Just give the manifest a temporary id to test storage API:
"browser_specific_settings": { "gecko": { "id": "extensionname@example.org", "strict_min_version": "88.0" } }

**Comment by Aayush9029** [comment]
Same response.

I updated the manifest.json with 
```json
  "browser_specific_settings": {
    "gecko": {
      "id": "extensionname@example.org",
      "strict_min_version": "88.0"
    }
  },  "browser_specific_settings": {
    "gecko": {
      "id": "extensionname@example.org",
      "strict_min_version": "88.0"
    }
  },
```
and ran like before
```bash
xcrun safari-web-extension-converter --copy-resources --swift --macos-only --app-name DuoOSU --project-location ~/Developer --bundle-identifier com.aayush.opensource.duoosufix FIREFOXCODEADDED
```

https://github.com/FreshSupaSulley/DuOSU/assets/43297314/e0ca7f9a-bcac-40e4-a752-ac24cb006538

**Comment by Aayush9029** [comment]
Sorry I haven't had time to go in-depth with your JS code. Do you think you can add an export option to export the secret as well as the counter.

iirc it's HOTP, unless they changed it later on.

**Comment by Aayush9029** [comment]
> how collaboration would work between us (and if that involves Apple Developer teams), signing, etc... I imagine we will need to publish it under your name (which is fine)? If you have any insight on the details that would be great.

Should we manage to make this work (assuming Apple will accept this extension), I would be more than happy to manage a Safari fork of the extension under an MIT license and to publish it for free.

**Comment by FreshSupaSulley** [comment]
IIRC, the export button exports literally everything put into the browser storage by DuOSU. There is no secret and there is no counter btw, DuOSU doesn't use a HOTP system. It stores keys, and then uses those credentials in the API calls to approve push requests. It's a bit different from other Duo extensions.

This is kind of annoying to remote debug. In both Chrome and Firefox, you can open an Inspector to analyze what's going wrong with the extension. Not sure if you can do the same in Safari.

And yes, please create a fork of your code. I'd love to help you figure this out, as I'm pretty sure you can still build extensions without the developer license, you just can't publish them.

**Comment by Aayush9029** [comment]
>ALSO, are you basing your code off of the Chrome branch or the Firefox branch?

I tried to chrome branch at first which was extremely buggy, Firefox worked better.

I'll take an in-depth look at this later this week. Thank you for your prompt response!

**Comment by Aayush9029** [comment]
If you or anyone reading this thread, has access to macOS device.

Here's macOS app archive.
*You might need to remove from quarantine as I didn't bother notarizing*

[DuoOSU.app.zip](https://github.com/FreshSupaSulley/DuOSU/files/12745528/DuoOSU.app.zip)

**Comment by FreshSupaSulley** [comment]
I also use a Mac. I'll take a look at this tomorrow 👍

**Comment by FreshSupaSulley** [comment]
I got DuOSU working for Safari without issues.
https://github.com/FreshSupaSulley/DuOSU/assets/45902499/8f49e387-44b6-4f33-b4fb-285d3d8f35de

It's definitely easier when you're the one who knows the code lmao.

**Comment by Aayush9029** [comment]
Wooo!

**Comment by FreshSupaSulley** [comment]
Almost. There was a lot of Safari CSS quirks, but the functionality worked just fine. There is still the issue of Xcode signing and ![what the hell the other files in the project mean](https://github.com/FreshSupaSulley/DuOSU/assets/45902499/4744f1a2-0855-487e-a19e-b6ea774e2970). Still gotta figure it out, but I'll make a fork for this soon.

**Comment by Aayush9029** [comment]
If you're referring to the red, outlined one then that's the DuOSU app itself.

**Comment by FreshSupaSulley** [comment]
Yeah, the red I don't know what it means. The DuOSU Extension part (green) is literally just the firefox code ported to Safari

**Comment by Aayush9029** [comment]
It’s technically “useless”; it merely meets Apple's requirement to have an app for Safari extensions. If you've used a Safari extension on iOS or macOS, you know it necessitates downloading an app and enabling it—it’s essentially the app placeholder.

**Comment by FreshSupaSulley** [comment]
Ok cool. I've never used Safari with extensions before my bad. I'll give the App part of the project some pretty design and replace the boilerplate icons and prompts.

**Comment by Aayush9029** [comment]
Great! To simplify matters, I'll notarize the macOS app, allowing you to do whatever you want with it. This will enable users to install it by just dragging and dropping, like any other app

**Comment by FreshSupaSulley** [comment]
Seems you have a lot of experience with Apple development!
... should we make the Xcode project a git repo?

Also how is publishing going to work, and what happens if we need to make a patch / update to it? Would I need to rely on you updating the app, or is there some sort of Apple Teams collab thing we can do?

**Comment by Aayush9029** [comment]
Also how is publishing going to work, and what happens if we need to make a patch / update to it?

Yeah you basically have to submit a new version of the app to the App Store, wait for the team App Store team to test and approve it and then publish it.

Although idk how the testers will test this.. I think the better option is to publish on your GitHub releases

**Comment by FreshSupaSulley** [comment]
If we don't publish it on the App store, wouldn't the user be required to allow unsigned extensions?
Or is that not how it works...? Like we can publish it **signed** by your developer license and then put it on Github?

**Comment by Aayush9029** [comment]
> If we don't publish it on the App store, wouldn't the user be required to allow unsigned extensions?

**Notarization works!**

Think of it like installing Firefox or chrome, is it on the App Store no? Can users install it without any issues yes.

**Comment by FreshSupaSulley** [comment]
Awesome... I was worried we would need to deal with the App Store bs. Didn't know that was an option.
I'm gonna fork what I have in the repo so you can take a look at it. I assume I'll just create the repo at the root of the Xcode project

**Comment by FreshSupaSulley** [comment]
The safari folder on the main branch has my Xcode project

**Comment by FreshSupaSulley** [comment]
btw - if you'd rather make a fork so we can work out of a separate repo we can do that too

**Comment by FreshSupaSulley** [comment]
(... does the thumbs down mean keep it all in this repo?)
I'm 90% sure the project is ready to be built. I assume I let you handle exporting the app with your Developer ID certificate and then we'll put it in the repo. Feel free to make any tweaks / credit yourself as you please.

**Comment by Aayush9029** [comment]
> (... does the thumbs down mean keep it all in this repo?)
Yes! If you really want to organize things, you can create a separate branch and call it "safari" 

Or this folder structure works too!

I'll work on this later this week!

**Comment by Aayush9029** [comment]
[DuOSU.app.zip](https://github.com/FreshSupaSulley/DuOSU/files/12863143/DuOSU.app.zip)
![CleanShot 10-10 at 20 36 49](https://github.com/FreshSupaSulley/DuOSU/assets/43297314/10170bb0-0ec9-45ac-a63c-84e1da056835)

**Comment by FreshSupaSulley** [comment]
You are the goat

... on launching the build, it doesn't appear in my extensions. So it looks like I still need to have Allow Unsigned Extensions enabled to get the binary to appear in Safari extensions. Is that the same for you? And if that's the case, we have a bit of an issue because Safari resets this every time it's closed. If not, there might be something wrong with my end?

In the screenshot, it looks like you changed the background to be translucent for the binary build (looks good), but I'm not seeing that on my end. So my first thought was that maybe my Safari is just using an old build I made for it a while ago, but that's probably not the case as I uninstalled my old DuOSU builds, restarted my Mac twice, and reinstalled your build.

I know nothing about notorizing. Is it possible that the app got notarized, but Safari still doesn't trust the extension embedded within it?
